packages/backend/src/open_payments/invoice/service.ts

```diff
@@ -110,62 +113,146 @@ async function handleInvoicePayment(
   if (!amountReceived) {
     return
   }
-  const invoice = await Invoice.query(deps.knex)
+  await Invoice.query(deps.knex)
     .patch({
-      active: false
+      active: false,
+      webhookId: uuid()
     })
     .where('id', invoiceId)
     .andWhere('amount', '<=', amountReceived.toString())
-    .returning('*')
-    .first()
-  if (invoice) {
-    await deps.webhookService.send({
-      type: EventType.InvoicePaid,
-      invoice,
-      amountReceived
-    })
-  }
 }
 
-// Deactivate expired invoices that have some money.
-// Delete expired invoices that have never received money.
+// Fetch (and lock) an invoice for work.
 // Returns the id of the processed invoice (if any).
-async function deactivateNextInvoice(
-  deps: ServiceDependencies
+async function processNextInvoice(
+  deps_: ServiceDependencies
 ): Promise<string | undefined> {
-  return deps.knex.transaction(async (trx) => {
+  return deps_.knex.transaction(async (trx) => {
+    const now = new Date(Date.now()).toISOString()
     // 30 seconds backwards to allow a prepared (but not yet fulfilled/rejected) packet to finish before being deactivated.
-    const now = new Date(Date.now() - 30_000).toISOString()
+    const delayedNow = new Date(Date.now() - 30_000).toISOString()
     const invoices = await Invoice.query(trx)
       .limit(1)
       // Ensure the invoices cannot be processed concurrently by multiple workers.
       .forUpdate()
       // If an invoice is locked, don't wait — just come back for it later.
       .skipLocked()
-      .where('active', true)
-      .andWhere('expiresAt', '<', now)
+      .where((builder: QueryBuilder) => {
+        builder.where('active', true).andWhere('expiresAt', '<', delayedNow)
+      })
+      .orWhere((builder: QueryBuilder) => {
+        builder.whereNotNull('webhookId').andWhere((builder: QueryBuilder) => {
+          builder
+            .where('webhookAttempts', 0)
+            // Back off between retries.
+            .orWhereRaw(
+              '"updatedAt" + LEAST("webhookAttempts", 6) * ? * interval \'1 seconds\' < ?',
+              [RETRY_BACKOFF_SECONDS, now]
+            )
+        })
```

Comment:
How can we ensure that this query executes efficiently?

I know that this is inspired by `OutgoingPayment`'s job queue, but in that case it relies on the index of the payment `state` and the fact that at any given time, most payments are either completed or cancelled.

Can we rely on `active` being `true` here, or not? If so then indexing on that would get a similar benefit as `state`. If not then we might have to store a `nextAttempt` timestamp, and index on that (rather than using `updatedAt` & `attempts`).